habitat_code/nobodyliveshere-kmeans.py

```diff
@@ -0,0 +1,804 @@
+# 
+# In the cell below, write a 2-3 sentence **site description** of this area that helps to put your analysis in context.
+
+# %% [markdown]
+# **YOUR SITE DESCRIPTION HERE**
+# 
+# ![](https://media.baamboozle.com/uploads/images/510741/1651543763_75056_gif-url.gif)
+
+# %% [markdown]
+# The watershed near New Orleans and the Mississippi Delta, marked by its low elevation and subtropical climate, is a critical ecological area where freshwater and marine environments merge. This region, with its unique blend of wetlands and estuaries, is shaped by the interplay of riverine inputs and oceanic influences, fostering a diverse ecosystem adapted to its warm, humid conditions.
+
+# %% [markdown]
+# ## STEP 3: MULTISPECTRAL DATA
+
+# %% [markdown]
+# ### Search for data
+# 
+# **YOUR TASK:**
+#   1. Log in to the `earthaccess` service using your Earthdata credentials:
+#      ```python
+#      earthaccess.login(persist=True)
+#      ```
+#   2. Modify the following sample code to search for granules of the HLSL30 product overlapping the watershed boundary from May to October 2023 (there should be 76 granules):
+#      ```python
+#      results = earthaccess.search_data(
+#          short_name="...",
+#          cloud_hosted=True,
+#          bounding_box=tuple(gruped_granule_grouped_granule_gdf.total_bounds),
+#          temporal=("...", "..."),
+#      )
+#      ```
+
+# %%
+eta.login
+
+results = eta.search_data(
+         short_name="HLSL30",
+         cloud_hosted=True,
+         bounding_box=tuple(watershed.total_bounds),
+         temporal=("2023-05-01", "2023-10-01"),
+    )
+
+# %% [markdown]
+# ### Compile information about each granule
+# 
+# I recommend building a GeoDataFrame, as this will allow you to plot the granules you are downloading and make sure they line up with your shapefile. You could also use a DataFrame, dictionary, or a custom object to store this information.
+# 
+# **YOUR TASK -- Write a function that:**
+#   1. For each search result:
+#       1. Get the following information (HINT: look at the ['umm'] values for each search result):
+#           - granule id (UR)
+#           - datetime
+#           - geometry (HINT: check out the shapely.geometry.Polygon class to convert points to a Polygon)
+#       2. Open the granule files. I recomment opening one granule at a time, e.g. with (`earthaccess.open([result]`).
+#       3. For each file (band), get the following information:
+#           - file handler returned from `earthaccess.open()`
+#           - tile id
+#           - band number
+#   2. Compile all the information you collected into a GeoDataFrame
+
+# %%
+import geopandas as gpd
+from shapely.geometry import Polygon
+import re
+
+def compile_granule_info(search_results):
+    granules_info = []
+    desired_bands = ['01','02', '03', '04', '05', '06', '07', '09']  # Bands to include

```

Comment:
Add comments to describe these bands. Alternatively, use a dictionary to map key:value pairs.

Reply:
Yeah, I should have gone back and clarified what the code was doing at certain points.  I think in this scenario I was having issues with the downloads so I was tinkering with how many bands I could fit without blowing up the file size.